File: go/indicators/johnehlers/hilberttransformer/estimator.go
package hilberttransformer

import "fmt"

const (
	defaultMinPeriod = 6
	defaultMaxPeriod = 50
	htLength         = 7
	quadratureIndex  = htLength / 2
)

// NewCycleEstimator creates a new cycle estimator based on the specified type and parameters.
func NewCycleEstimator(typ CycleEstimatorType, params *CycleEstimatorParams) (CycleEstimator, error) {
	switch typ {
	case HomodyneDiscriminator:
		return NewHomodyneDiscriminatorEstimator(params)
	case HomodyneDiscriminatorUnrolled:
		return NewHomodyneDiscriminatorEstimatorUnrolled(params)
	case PhaseAccumulator:
		return NewPhaseAccumulatorEstimator(params)
	case DualDifferentiator:
		return NewDualDifferentiatorEstimator(params)
	}

	return nil, fmt.Errorf("invalid cycle estimator type: %s", typ)
}

// EstimatorMoniker returns the moniker of the cycle estimator.
func EstimatorMoniker(typ CycleEstimatorType, estimator CycleEstimator) string {
	namer := func(s string, e CycleEstimator) string {
		const f = "%s(%d, %.3f, %.3f)"

		return fmt.Sprintf(f, s, e.SmoothingLength(), e.AlphaEmaQuadratureInPhase(), e.AlphaEmaPeriod())
	}

	switch typ {
	case HomodyneDiscriminator:
		return namer("hd", estimator)
	case HomodyneDiscriminatorUnrolled:
		return namer("hdu", estimator)
	case PhaseAccumulator:
		return namer("pa", estimator)
	case DualDifferentiator:
		return namer("dd", estimator)
	}

	return ""
}

// Push shifts all elements to the right and place the new value at index zero.
func push(array []float64, value float64) {
	for i := len(array) - 1; i > 0; i-- {
		array[i] = array[i-1]
	}

	array[0] = value
}

func correctAmplitude(previousPeriod float64) float64 {
	const (
		a = 0.54
		b = 0.075
	)

	return a + b*previousPeriod
}

func ht(array []float64) float64 {
	const (
		a = 0.0962
		b = 0.5769
	)

	value := 0.
	value += a * array[0]
	value += b * array[2]
	value -= b * array[4]
	value -= a * array[6]

	return value
}

func adjustPeriod(period, periodPrevious float64) float64 {
	const (
		minPreviousPeriodFactor = 0.67
		maxPreviousPeriodFactor = 1.5
	)

	temp := maxPreviousPeriodFactor * periodPrevious
	if period > temp {
		period = temp
	} else {
		temp = minPreviousPeriodFactor * periodPrevious
		if period < temp {
			period = temp
		}
	}

	if period < defaultMinPeriod {
		period = defaultMinPeriod
	} else if period > defaultMaxPeriod {
		period = defaultMaxPeriod
	}

	return period
}

func fillWmaFactors(length int, factors []float64) {
	const (
		c4   = 4
		c3   = 3
		c410 = 4. / 10.
		c310 = 3. / 10.
		c210 = 2. / 10.
		c110 = 1. / 10.
		c36  = 3. / 6.
		c26  = 2. / 6.
		c16  = 1. / 6.
		c23  = 2. / 3.
		c13  = 1. / 3.
	)

	if length == c4 {
		factors[0] = c410
		factors[1] = c310
		factors[2] = c210
		factors[3] = c110
	} else if length == c3 {
		factors[0] = c36
		factors[1] = c26
		factors[2] = c16
	} else { // if length == 2
		factors[0] = c23
		factors[1] = c13
	}
}

func verifyParameters(p *CycleEstimatorParams) error {
	const (
		invalid = "invalid cycle estimator parameters"
		fmts    = "%s: %s"
		minLen  = 2
		maxLen  = 4
	)

	length := p.SmoothingLength
	if length < minLen || length > maxLen {
		return fmt.Errorf(fmts, invalid, "SmoothingLength should be in range [2, 4]")
	}

	alphaQuad := p.AlphaEmaQuadratureInPhase
	if alphaQuad <= 0 || alphaQuad >= 1 {
		return fmt.Errorf(fmts, invalid, "AlphaEmaQuadratureInPhase should be in range (0, 1)")
	}

	alphaPeriod := p.AlphaEmaPeriod
	if alphaPeriod <= 0 || alphaPeriod >= 1 {
		return fmt.Errorf(fmts, invalid, "AlphaEmaPeriod should be in range (0, 1)")
	}

	return nil
}

File: py/indicators/john_ehlers/hilbert_transformer/estimator.py
"""Shared helper functions for Hilbert transformer cycle estimators."""
import math

from .cycle_estimator import CycleEstimator
from .cycle_estimator_params import CycleEstimatorParams
from .cycle_estimator_type import CycleEstimatorType

DEFAULT_MIN_PERIOD = 6
DEFAULT_MAX_PERIOD = 50
HT_LENGTH = 7
QUADRATURE_INDEX = HT_LENGTH // 2
ACCUMULATION_LENGTH = 40


def push(array: list[float], value: float) -> None:
    """Shifts all elements to the right and places the new value at index zero."""
    for i in range(len(array) - 1, 0, -1):
        array[i] = array[i - 1]
    array[0] = value


def ht(array: list[float]) -> float:
    """Hilbert transform of a 7-element array."""
    a = 0.0962
    b = 0.5769
    return a * array[0] + b * array[2] - b * array[4] - a * array[6]


def correct_amplitude(previous_period: float) -> float:
    """Computes amplitude correction factor."""
    return 0.54 + 0.075 * previous_period


def adjust_period(period: float, period_previous: float) -> float:
    """Clamps period to [0.67x, 1.5x] of previous and [6, 50] absolute."""
    temp = 1.5 * period_previous
    if period > temp:
        period = temp
    else:
        temp = 0.67 * period_previous
        if period < temp:
            period = temp

    if period < DEFAULT_MIN_PERIOD:
        period = DEFAULT_MIN_PERIOD
    elif period > DEFAULT_MAX_PERIOD:
        period = DEFAULT_MAX_PERIOD

    return period


def fill_wma_factors(length: int, factors: list[float]) -> None:
    """Fills WMA weight factors for the given length (2, 3, or 4)."""
    if length == 4:
        factors[0] = 4.0 / 10.0
        factors[1] = 3.0 / 10.0
        factors[2] = 2.0 / 10.0
        factors[3] = 1.0 / 10.0
    elif length == 3:
        factors[0] = 3.0 / 6.0
        factors[1] = 2.0 / 6.0
        factors[2] = 1.0 / 6.0
    else:  # length == 2
        factors[0] = 2.0 / 3.0
        factors[1] = 1.0 / 3.0


def verify_parameters(p: CycleEstimatorParams) -> None:
    """Validates cycle estimator parameters. Raises ValueError if invalid."""
    invalid = "invalid cycle estimator parameters"

    if p.smoothing_length < 2 or p.smoothing_length > 4:
        raise ValueError(f"{invalid}: SmoothingLength should be in range [2, 4]")

    if p.alpha_ema_quadrature_in_phase <= 0 or p.alpha_ema_quadrature_in_phase >= 1:
        raise ValueError(
            f"{invalid}: AlphaEmaQuadratureInPhase should be in range (0, 1)"
        )

    if p.alpha_ema_period <= 0 or p.alpha_ema_period >= 1:
        raise ValueError(f"{invalid}: AlphaEmaPeriod should be in range (0, 1)")


def estimator_moniker(typ: CycleEstimatorType, estimator: CycleEstimator) -> str:
    """Returns the moniker of the cycle estimator."""
    prefixes = {
        CycleEstimatorType.HOMODYNE_DISCRIMINATOR: "hd",
        CycleEstimatorType.HOMODYNE_DISCRIMINATOR_UNROLLED: "hdu",
        CycleEstimatorType.PHASE_ACCUMULATOR: "pa",
        CycleEstimatorType.DUAL_DIFFERENTIATOR: "dd",
    }

    prefix = prefixes.get(typ)
    if prefix is None:
        return ""

    return f"{prefix}({estimator.smoothing_length()}, " \
           f"{estimator.alpha_ema_quadrature_in_phase():.3f}, " \
           f"{estimator.alpha_ema_period():.3f})"


def new_cycle_estimator(
    typ: CycleEstimatorType, params: CycleEstimatorParams
) -> CycleEstimator:
    """Creates a new cycle estimator based on the specified type and parameters."""
    from .homodyne_discriminator_estimator import HomodyneDiscriminatorEstimator
    from .homodyne_discriminator_estimator_unrolled import \
        HomodyneDiscriminatorEstimatorUnrolled
    from .phase_accumulator_estimator import PhaseAccumulatorEstimator
    from .dual_differentiator_estimator import DualDifferentiatorEstimator

    if typ == CycleEstimatorType.HOMODYNE_DISCRIMINATOR:
        return HomodyneDiscriminatorEstimator(params)
    elif typ == CycleEstimatorType.HOMODYNE_DISCRIMINATOR_UNROLLED:
        return HomodyneDiscriminatorEstimatorUnrolled(params)
    elif typ == CycleEstimatorType.PHASE_ACCUMULATOR:
        return PhaseAccumulatorEstimator(params)
    elif typ == CycleEstimatorType.DUAL_DIFFERENTIATOR:
        return DualDifferentiatorEstimator(params)
    else:
        raise ValueError(f"invalid cycle estimator type: {typ}")
